### DISCUSSION TITLE
Display local time based on location of each connected user

### DISCUSSION THREAD

**Discussion by jose1711** [jamulussoftware/jamulus#2341] [Ideas]
It could be nice if Jamulus shows local time for each connected user based on his/her country/location. Rough idea:

![obrázok](https://user-images.githubusercontent.com/1406222/152514376-ca4b65af-257a-4a0d-85f3-2a8294dd09d7.png)

**Comment by pljones** [comment]
This could be done roughly in the client based on the flag look up to locale to get offset from the current locale, I guess...

What should happen if a user hasn't set their country?  No time, UTC, local time?

**Reply by rdica** [reply]
If such a thing were to be considered, I'd expect it to be a setting thats off by default. If I wanted others to know what TZ I'm in, I'll tell them. Same as setting city and country currently in your profile.

**Reply by jose1711** [reply]
I agree with @rdica - stick to strict privacy settings in the defaults. And off could perhaps be shown as `??:??` or `--:--`. The display of TZ could also be toggled (I can imagine some people are just happy with the information Jamulus is showing already)

**Reply by dcorson-ticino-com** [reply]
I have always simply talked with the others I'm playing with.
That has always worked fine.

**Reply by pljones** [reply]
There would be no new information exchanged if it were done the same way as the flag.  What I suggested is exactly what everyone can do now - look at the flag, find the likely timezone, calculate the time there.

It would be down to each client whether to display the local time of each channel, of course.

**Reply by dcorson-ticino-com** [reply]
Sure it should be simple to do and not cause any additional network traffic. I just think it will be, however, unnecessary GUI clutter, something moving, continuously updating with little added information.  I would have less problems imagining a static display like GMT+1 or whatever corresponding to the locale.
But even then, the user info area is already very dense and not always clear, I would not like to add more to it.  
But if we can find a way to make it look good without gobbling lots of screen real estate I am not opposed.

**Reply by jose1711** [reply]
> Sure it should be simple to do and not cause any additional network traffic.

Don't think there will be any penalty caused  by additional network traffic. The timezone can be communicated once - when the new clients connects. There is no need to keep the time in sync as the user timezone can be calculated from the local time.

**Reply by jose1711** [reply]
> There would be no new information exchanged if it were done the same way as the flag. What I suggested is exactly what everyone can do now - look at the flag, find the likely timezone, calculate the time there.

How about the fact Russia spans across 11 timezones? That would have to be a very wild guess if only flag is used for calculation.

**Reply by jose1711** [reply]
> But even then, the user info area is already very dense and not always clear, I would not like to add more to it. But if we can find a way to make it look good without gobbling lots of screen real estate I am not opposed.

Then perhaps a mouse-over tooltip would be a better place? And speaking of clutter.. I think the tooltips could be reorganized a little bit. They seem to be unnecessarily narrow to me (but that's a bit OT).

**Reply by pljones** [reply]
> The timezone can be communicated once 

No, it won't be added to the protocol messages.  That causes client incompatibility.  There's no need to change the protocol for this.

**Reply by jose1711** [reply]
> > The timezone can be communicated once
> 
> No, it won't be added to the protocol messages. That causes client incompatibility. There's no need to change the protocol for this.

Sorry, I am not really familiar with the protocol internals. Just wanted to say there's probably no need to send this info more than once (so the network overhead should be negligible).

**Reply by pljones** [reply]
I'm just saying the timezone won't be communicated _at all_.  The existing user information will be used to derive an approximately correct value.  It's not sufficiently important to do otherwise.

**Reply by dcorson-ticino-com** [reply]
How to do without changing protocol ???
Database with all countries and their cities with timezone ???
The solution is not obvious to me.
I'd say, do so it is right or not at all, a wrong timezone is more confusing than no timezone.

**Reply by pljones** [reply]
Every client already has a location (Qt Locale) for every channel.  That's enough for the _client_ to determine an approximate timezone.  If the Qt Locale is Brazil, you'll get Brasilia Standard Time - might not be perfect but...

**Reply by dcorson-ticino-com** [reply]
> Every client already has a location (Qt Locale) for every channel. That's enough for the _client_ to determine an approximate timezone. If the Qt Locale is Brazil, you'll get Brasilia Standard Time - might not be perfect but...

Does the Locale differentiate if the client is on the, for example, US east or west coast ?  I am 6hours from EST but 9 hours from PST. 
If we can not get that right it is pretty useless and I don't think we should do it.

**Reply by pljones** [reply]
Like I said, appoximate: Brazil has four timezones.  That's the best that's going to be done without breaking things.

**Reply by dcorson-ticino-com** [reply]
> Like I said, appoximate: Brazil has four timezones. That's the best that's going to be done without breaking things.

In that case it is less than useless.

**Comment by pljones** [comment]
> I just think it will be, however, unnecessary GUI clutter, something moving, continuously updating with little added information.

I agree.  But It doesn't mean _anyone_ will be _forced_ to use the feature.
